MRG: implementing / testing get_fdata

matthew-brett opened the pull request:
get_fdata the new proposed default method for getting the data from the image.
It always returns floating point.

effigies commented on nibabel/dataobj_images.py:
```diff
@@ -191,11 +204,155 @@ def get_data(self, caching='fill'):
+        >>> data_again[0, 0, 0, 0]
+        0.0
+        """
+        if caching not in ('fill', 'unchanged'):
```
Want to test `caching='unknown'`?

satra commented on nibabel/dataobj_images.py:
```diff
@@ -191,11 +204,155 @@ def get_data(self, caching='fill'):
             self._data_cache = data
         return data
 
+    def get_fdata(self, caching='fill', dtype=np.float64):
```
any reason why float64 over float32 is the default?

matthew-brett replied:
@satra - the default up till now has been to return float64 for images with scaling.  Also, some images do have float64 and it seems a shame to downgrade them with this call.  So float32 seems to me like an option rather than a default.

satra commented:
@matthew-brett - other than my question, which should prevent this from being merged, and on top of @effigies review, the code and documentation looks good.

effigies commented:
Shouldn't, maybe?

satra commented:
thanks @effigies - that was definitely a lapso

matthew-brett commented:
Guys - thanks for the reviews.  I've updated to respond to your comments, and I've added some more tests of the caching, along with a fix to the code.  To help reviewing, can I suggest that y'all look at commit c729e06 separately (it has the caching changes, and tests)?

matthew-brett commented:
Oops sorry - I missed a changed I needed to make in the tests, to reflect the change in caching behavior.  I rebased.  New commit with all the relevant changes is c800493.

satra commented:
+1 from me as well.

matthew-brett commented:
@effigies - any interests in being release manager again?  It's ready as far as I am concerned - but I've got 15 hours of completely new teaching to prepare in the next couple of weeks, and I'm really slammed for time.

effigies commented:
Sure. I'll aim to release by the end of the week. I'll let you know if I have any issues.

mwaskom commented:
Was this discussed more somewhere else? The release notes link here for information about this change, but there's no information in the PR (aside from the code itself!) that motivates the new function. (I ask as it's my vague sense that `get_fdata` is intended to replace `get_data`).

effigies commented:
@mwaskom This thread is probably the better link... https://mail.python.org/pipermail/neuroimaging/2017-April/001383.html

mwaskom commented:
Thanks @effigies. Perhaps the PR intro could be edited to include this link.

effigies commented:
Done.
